### PR TITLE
refactor Create or update and items bug fixes

### PR DESCRIPTION
**Pull request by lauherkcambria**
#### What tests were added (if relevant), and do all tests (rspec, karma, and integration CALS tests) pass?
- N/A as its a refactor

#### Which User story does this relate to?
-https://www.pivotaltracker.com/n/projects/1662141/stories/150103623
-https://www.pivotaltracker.com/n/projects/1662141/stories/150103559

#### Brief feature description
- cleans up controller
- fixes create or update method
- fixes items problem
- fixes some bad data coming into rails side
- whitelists params

#### Does this PR block/effect other stories?
- https://www.pivotaltracker.com/n/projects/1662141/stories/150111734

#### Do any commands needs to be run in order to run the project with this PR?
- na

#### Code coverage percent diff as a result of this PR?
- na

#### JS errors diff as a result of this PR?
-na

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -16,7 +16,7 @@ gem 'responders', '~> 2.3'
 gem 'autoprefixer-rails', '~> 6.7'
 gem 'bootstrap-sass', '~> 3.3.6'
 gem 'bootstrap'
-
+gem 'awesome_print', :require => 'ap'
```
Comment (ratneshraval) [inline Gemfile]: Do we need this gem in prod?

File: app/javascript/rfa_forms/rfa01a_edit_view/minorCardField.js [hunk 2]
```diff
@@ -20,7 +20,7 @@ export class MinorCardField extends React.Component {
           selectClassName='reusable-select'
           optionList={setToWhom(this.props.applicants)}
           label='To whom'
-          onChange={(event) => this.props.handleRelationshipTypeToApplicant(this.props.index, event.target.value, 'applicant_id')} />
+          onChange={(event) => this.props.handleToWhom(this.props.index, event.target.value, this.props.applicants)} />
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/minorCardField.js]: The development branch is correct, this overrides the code with older version.

File: app/models/helpers/dictionary.rb [hunk 3]
```diff
@@ -19,12 +19,58 @@ def residence_ownership_types
...
+
   def license_types
     Dictionaries::LicenseType.all(auth_header)
   end
 
   def relationship_to_applicant_types
     Dictionaries::RelationshipToApplicantType.all(auth_header)
   end
+
+  def  name_types
```
Comment (ratneshraval) [inline app/models/helpers/dictionary.rb]: 👍 Good work on combining all rfa dictionaries

File: app/javascript/rfa_forms/rfa01a_edit_view/OtherAdultsCardsGroup.js [hunk 4]
```diff
@@ -75,17 +75,17 @@ export default class OtherAdultsCardsGroup extends React.Component {
     let relationshipList = Immutable.fromJS(otherAdults[index].relationship_to_applicants[0])
 
     relationshipList = relationshipList.setIn(['relationship_to_applicant', 'id'], value)
-    relationshipList = relationshipList.setIn(['relationship_to_applicant', 'value'], this.props.relationship_types[value].value)
+    relationshipList = relationshipList.setIn(['relationship_to_applicant', 'value'], this.props.relationship_types[value - 1].value)
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/OtherAdultsCardsGroup.js]: We shouldn't trying to match value by subtracting here. We need to use the ID sent by api.

File: Gemfile (hunk 1 above)
Comment (lauherkcambria) [inline Gemfile]: good catch, i'll move to dev

File: app/models/helpers/dictionary.rb (hunk 3 above)
Comment (lauherkcambria) [inline app/models/helpers/dictionary.rb]: thanks 😄

File: app/javascript/rfa_forms/rfa01a_edit_view/OtherAdultsCardsGroup.js (hunk 4 above)
Comment (lauherkcambria) [inline app/javascript/rfa_forms/rfa01a_edit_view/OtherAdultsCardsGroup.js]: i agree -- i'll fix this in the validate other adults fields story

File: app/controllers/rfa/a01_controller.rb [hunk 5]
```diff
@@ -7,114 +7,133 @@ def index
...
-      adoption_history_results = rfa_adoption_history_helper.create(params[:id], adoption_history.to_json)
-      application_response[:adoption_history] = adoption_history_results
+    def adoption_history_params
+      params.require(:fosterCareHistory).permit(:id, :was_subject_for_exclusion_order_q7,
+      :foster_care_licenses_q1 =>[:was_previously_licensed, :agencies =>[:name, :type=>[:id, :value]]],
+      :applications_for_adoption_q2 => [:have_applied_for_adoption, :facilities => []],
+      :facility_operation_licenses_q3 =>[:was_previously_licensed, :agencies =>[:name, :type=>[:id, :value]]],
+      :employment_in_facilities_q4 => [:was_employed_or_volunteered, :facilities => []],
+      :denial_history_q5 =>[:had_denials, :agencies =>[:name, :type=>[:id, :value]]],
+      :suspension_revocation_history_q6 =>[:had_suspensions_revocations, :agencies =>[:name, :type=>[:id, :value]]])
```
Comment (lavanyareddyreddy) [inline app/controllers/rfa/a01_controller.rb]: @lauherkcambria : The reason why I didn't add white lists to strong Params initially is because I didn't wanted to overload controller but now that we are adding whitelist, extracting params logic into a class will make controller thinner, we will be able to reuse it and it will be easier to test. 
Here is an example http://railscasts.com/episodes/371-strong-parameters

File: app/models/concerns/rfa_base_api_protocol_provider.rb [hunk 6]
```diff
@@ -4,13 +4,26 @@ module Concerns::RfaBaseApiProtocolProvider
 
   class_methods do
     def create(auth_header, parent_id, body)
+puts body
```
Comment (lavanyareddyreddy) [inline app/models/concerns/rfa_base_api_protocol_provider.rb]: @lauherkcambria : can you remove the puts

File: app/controllers/rfa/a01_controller.rb [hunk 7]
```diff
@@ -7,114 +7,133 @@ def index
...
+    end
 
-  private
+    def create_or_update(item, helper, parent_id)
+      if item[:id]
+        result = helper.update(parent_id, item[:id], item.to_json)
+      else
+        result = helper.create(parent_id, item.to_json)
+      end
+      return result
```
Comment (lauherkcambria) [inline app/controllers/rfa/a01_controller.rb]: you're right, it doesn't matter. I prefer to leave it because it's totally style.

File: app/models/concerns/rfa_base_api_protocol_provider.rb (hunk 6 above)
Comment (lauherkcambria) [inline app/models/concerns/rfa_base_api_protocol_provider.rb]: yes

File: app/controllers/rfa/a01_controller.rb (hunk 5 above)
Comment (lauherkcambria) [inline app/controllers/rfa/a01_controller.rb]: the controller is much less overloaded now than it was before (code climate went from C -> A with this refactor), and i'm . following the guidelines for rails 5 strong paramenters,  [link](http://guides.rubyonrails.org/action_controller_overview.html#strong-parameters). I'm not sure about the link you provided (it's from 2012 and rails 3),   but Sure, you can do that refactor when it's used in minor children card and we can review then.

File: app/controllers/rfa/a01_controller.rb (hunk 7 above)
Comment (lavanyareddyreddy) [inline app/controllers/rfa/a01_controller.rb]: Its not just style I think because its to do with maintainability, keeping controller cleaner/thin and reusability.

File: app/javascript/rfa_forms/rfa01a_edit_view/minorCardsGroup.jsx [hunk 8]
```diff
@@ -34,14 +43,37 @@ export default class MinorCardsGroup extends React.Component {
   }
 
   onFieldChange (cardIndex, value, type) {
-    let minorChildrenList = Immutable.fromJS(this.props.minorChildren)
+    let minorChildrenList = Immutable.fromJS(this.props.minorChildren || [minorDefaults])
```
Comment (lauherkcambria) [inline app/javascript/rfa_forms/rfa01a_edit_view/minorCardsGroup.jsx]: because none of it worked when trying to save data. all changes i made to minor children was in order to make it work from when i pulled development. Nesh asked me to revert these changes, however.

File: app/controllers/rfa/a01_controller.rb (hunk 7 above)
Comment (lauherkcambria) [inline app/controllers/rfa/a01_controller.rb]: i disagree. how is 'return result' less maintainable than 'result'...?
